File: invmass.cpp
#include <TFile.h>
#include <TTree.h>
#include <TH1F.h>
#include <TCanvas.h>
#include <TF1.h>
#include <TRandom3.h>
#include <Math/Vector4D.h>
#include <Math/Boost.h>
#include <vector>
#include <cmath>
#include <iostream>

using namespace ROOT::Math;
using namespace std;

// Smear particle momentum
PxPyPzMVector smear_momentum(const PxPyPzMVector &particle, TRandom3 &randGen) {
    double fluct = randGen.Gaus(1., 0.05);
    return PxPyPzMVector(particle.Px()*fluct, particle.Py()*fluct, particle.Pz()*fluct, particle.M());
}

// Generate two-body decay in parent rest frame
pair<PxPyPzMVector, PxPyPzMVector> two_body_decay(const PxPyPzMVector &parent, double m1, double m2, TRandom3 &randGen) {
    double m0 = parent.M();
    double p12 = sqrt((m0*m0 - pow(m1+m2,2))*(m0*m0 - pow(m1-m2,2)))/(2.*m0);
    double E1 = (m0*m0 + m1*m1 - m2*m2)/(2.*m0);
    double p = sqrt(E1*E1 - m1*m1);

    double theta = randGen.Uniform(0, M_PI);
    double phi = randGen.Uniform(0, 2*M_PI);

    double p1x = p*sin(theta)*cos(phi);
    double p1y = p*sin(theta)*sin(phi);
    double p1z = p*cos(theta);

    PxPyPzMVector daughter1(p1x, p1y, p1z, m1);
    PxPyPzMVector daughter2(-p1x, -p1y, -p1z, m2);

    Boost boost(parent.BoostToCM());
    daughter1 = boost(daughter1);
    daughter2 = boost(daughter2);

    return make_pair(daughter1, daughter2);
}

// Generate parent particle and decay into two daughters
vector<PxPyPzMVector> generate_and_decay(double parent_mass, double m1, double m2, TRandom3 &randGen) {
    double momentum = randGen.Exp(1.0);
    double theta = randGen.Uniform(0, M_PI);
    double phi = randGen.Uniform(0, 2*M_PI);

    double px = momentum*sin(theta)*cos(phi);
    double py = momentum*sin(theta)*sin(phi);
    double pz = momentum*cos(theta);

    PxPyPzMVector parent(px, py, pz, parent_mass);

    auto daughters = two_body_decay(parent, m1, m2, randGen);

    vector<PxPyPzMVector> tracks;
    tracks.push_back(smear_momentum(daughters.first, randGen));
    tracks.push_back(smear_momentum(daughters.second, randGen));

    return tracks;
}

int invmass() {
    gROOT->SetBatch(kTRUE);  // No GUI windows
    TRandom3 randGen(42);

    int nParticles = 1000;

    double mass_pi_ch = 0.13957;
    double mass_k_zero = 0.497611;
    double mass_d_zero = 1.86484;

    TH1F *hInvMass = new TH1F("hInvMass", "Invariant Mass", 300, 0, 3);

    TFile *fileout = new TFile("tracks.root", "RECREATE");
    TTree *tree = new TTree("tree", "Tree with tracks");
    vector<PxPyPzMVector> tracks_vec;
    tree->Branch("tracks", &tracks_vec);

    for(int i=0; i<nParticles; i++) {
        tracks_vec.clear();
        auto tracks1 = generate_and_decay(mass_k_zero, mass_pi_ch, mass_pi_ch, randGen);
        auto tracks2 = generate_and_decay(mass_d_zero, mass_pi_ch, mass_pi_ch, randGen);

        tracks_vec.insert(tracks_vec.end(), tracks1.begin(), tracks1.end());
        tracks_vec.insert(tracks_vec.end(), tracks2.begin(), tracks2.end());

        // Fill histogram
        for(size_t j=0; j<tracks_vec.size(); j++) {
            for(size_t k=j+1; k<tracks_vec.size(); k++) {
                hInvMass->Fill((tracks_vec[j]+tracks_vec[k]).M());
            }
        }

        tree->Fill();
    }

    fileout->cd();
    tree->Write();
    fileout->Close();

    // Fit function: two Gaussians + 5th-degree polynomial
    TF1 *fitFunc = new TF1("fitFunc",
                           "[0]/(sqrt(2*TMath::Pi())*[2])*TMath::Exp(-0.5*((x-[1])/[2])^2)"
                           "+ [3]/(sqrt(2*TMath::Pi())*[5])*TMath::Exp(-0.5*((x-[4])/[5])^2)"
                           "+ [6] + [7]*x + [8]*x^2 + [9]*x^3 + [10]*x^4 + [11]*x^5",
                           -10, 10);

    Double_t params[12] = {1, 0.5, 0.01, 1, 1.85, 0.05, 0, 0, 0, 0, 0, 0};
    fitFunc->SetParameters(params);

    fitFunc->SetParLimits(1, 0.48, 0.52); // K0 mass
    fitFunc->SetParLimits(2, 0.0, 0.1);   // K0 width
    fitFunc->SetParLimits(4, 1.8, 1.9);   // D0 mass
    fitFunc->SetParLimits(5, 0.0, 0.2);   // D0 width

    hInvMass->Fit(fitFunc);

    Double_t bin_width = hInvMass->GetBinLowEdge(2) - hInvMass->GetBinLowEdge(1);
    cout << "Number of signal events #1 = " << fitFunc->GetParameter(0)/bin_width
         << " +- " << fitFunc->GetParError(0)/bin_width << endl;
    cout << "Number of signal events #2 = " << fitFunc->GetParameter(3)/bin_width
         << " +- " << fitFunc->GetParError(3)/bin_width << endl;

    TCanvas *canvas = new TCanvas("canvas", "Invariant Mass", 600, 600);
    hInvMass->GetXaxis()->SetTitle("M(#pi^{+}#pi^{-}) [GeV]");
    hInvMass->GetYaxis()->SetTitle("Events");
    hInvMass->Draw();

    canvas->SaveAs("invmass.pdf");
    canvas->SaveAs("invmass.png");

    return 0;
}

File: invariant/invmass.py
import ROOT
import math
import random
import tqdm
import sys
from array import array

def generate_and_decay(parent_mass, daughter_mass_1, daughter_mass_2):
    # Generate random values for momentum and direction
    momentum = random.expovariate(1.0)
    theta = random.uniform(0, math.pi)
    phi = random.uniform(0, 2 * math.pi)

    px = momentum * math.sin(theta) * math.cos(phi)
    py = momentum * math.sin(theta) * math.sin(phi)
    pz = momentum * math.cos(theta)

    parent = ROOT.Math.PxPyPzMVector(px, py, pz, parent_mass)
    daughter1, daughter2 = two_body_decay(parent, daughter_mass_1, daughter_mass_2)

    # Smear momentum of daughter tracks
    daughter1 = smear_momentum(daughter1)
    daughter2 = smear_momentum(daughter2)

    return [daughter1, daughter2]

def smear_momentum(particle):
    fluct = random.normalvariate(1., 0.05)
    return ROOT.Math.PxPyPzMVector(particle.X() * fluct,
                                   particle.Y() * fluct,
                                   particle.Z() * fluct,
                                   particle.M())

def two_body_decay(parent, m1, m2):
    m0 = parent.M()
    p12 = math.sqrt(max(0.0, (m0**2 - (m1 + m2)**2) * (m0**2 - (m1 - m2)**2))) / (2 * m0)
    E1 = (m0**2 + m1**2 - m2**2) / (2 * m0)
    p = math.sqrt(max(0.0, E1**2 - m1**2))

    theta = random.uniform(0, math.pi)
    phi = random.uniform(0, 2 * math.pi)

    p1x = p * math.sin(theta) * math.cos(phi)
    p1y = p * math.sin(theta) * math.sin(phi)
    p1z = p * math.cos(theta)

    daughter1 = ROOT.Math.PxPyPzMVector(p1x, p1y, p1z, m1)
    daughter2 = ROOT.Math.PxPyPzMVector(-p1x, -p1y, -p1z, m2)

    boost_vector = ROOT.Math.Boost(parent.BoostToCM())
    daughter1 = ROOT.Math.PxPyPzMVector(boost_vector(daughter1))
    daughter2 = ROOT.Math.PxPyPzMVector(boost_vector(daughter2))

    return daughter1, daughter2

if __name__ == '__main__':
    nParticles = 1000
    mass_pi_ch = 0.13957
    mass_k_zero = 0.497611
    mass_d_zero = 1.86484
    mass_b_zero = 5.27958

    random.seed(42)
    hInvMass = ROOT.TH1F("hInvMass", "Invariant Mass", 300, 0, 6)

    fileout = ROOT.TFile("tracks.root", "recreate")
    tree = ROOT.TTree('tree', 'Tree with tracks')
    tracks_vec = ROOT.std.vector(ROOT.Math.PxPyPzMVector)()
    branch = tree.Branch('tracks', tracks_vec)

    for i in tqdm.tqdm(range(nParticles)):
        tracks = []
        tracks += generate_and_decay(mass_k_zero, mass_pi_ch, mass_pi_ch)
        tracks += generate_and_decay(mass_d_zero, mass_pi_ch, mass_pi_ch)
        tracks += generate_and_decay(mass_b_zero, mass_pi_ch, mass_pi_ch)
        assert len(tracks) == 6

        for itr1 in range(len(tracks)):
            for itr2 in range(itr1 + 1, len(tracks)):
                hInvMass.Fill((tracks[itr1] + tracks[itr2]).M())

        tracks_vec.clear()
        for tr in tracks:
            tracks_vec.push_back(tr)
        tree.Fill()

    fileout.cd()
    tree.Write()
    fileout.Close()

 
    fitFunc = ROOT.TF1("fitFunc",
                       "[0]/(sqrt(2*TMath::Pi())*[2])*TMath::Exp(-0.5*((x-[1])/[2])^2)"   # K0
                       " + [3]/(sqrt(2*TMath::Pi())*[5])*TMath::Exp(-0.5*((x-[4])/[5])^2)" # D0
                       " + [6]/(sqrt(2*TMath::Pi())*[8])*TMath::Exp(-0.5*((x-[7])/[8])^2)" # B0
                       " + [9] + [10]*x + [11]*x^2 + [12]*x^3 + [13]*x^4 + [14]*x^5",     # Polynomial
                       0, 6)

    params = array('d', [
        200., 0.4976, 0.02,   # K0
        500., 1.865, 0.03,    # D0
        300., 5.2796, 0.05,   # B0
        0, 0, 0, 0, 0, 0      # Polynomial
    ])
    fitFunc.SetParameters(params)

    fitFunc.SetParLimits(1, 0.48, 0.52)   # K0 mean
    fitFunc.SetParLimits(2, 0., 0.1)      # K0 sigma
    fitFunc.SetParLimits(4, 1.80, 1.90)   # D0 mean
    fitFunc.SetParLimits(5, 0., 0.2)      # D0 sigma
    fitFunc.SetParLimits(7, 5.20, 5.35)   # B0 mean
    fitFunc.SetParLimits(8, 0., 0.5)      # B0 sigma

    hInvMass.Fit(fitFunc)

    bin_width = hInvMass.GetBinWidth(1)
    print(f'Number of signal events K0 = {fitFunc.GetParameter(0)/bin_width:.1f} ± {fitFunc.GetParError(0)/bin_width:.1f}')
    print(f'Number of signal events D0 = {fitFunc.GetParameter(3)/bin_width:.1f} ± {fitFunc.GetParError(3)/bin_width:.1f}')
    print(f'Number of signal events B0 = {fitFunc.GetParameter(6)/bin_width:.1f} ± {fitFunc.GetParError(6)/bin_width:.1f}')

    canvas = ROOT.TCanvas("canvas", "Invariant Mass", 600, 600)
    hInvMass.GetXaxis().SetTitle('M(π⁺π⁻) [GeV]')
    hInvMass.GetYaxis().SetTitle('Events')
    hInvMass.Draw()
    fitFunc.Draw("same")

    canvas.SaveAs("invmass_1.pdf")
    canvas.SaveAs("invmass_1.png")
